src/notebooks/controllers/controllerDefaultService.ts

```diff
@@ -0,0 +1,125 @@
+import { InteractiveWindowView, JupyterNotebookView, PYTHON_LANGUAGE } from '../../platform/common/constants';
+import { IDisposableRegistry, IsWebExtension, Resource } from '../../platform/common/types';
+import { getNotebookMetadata } from '../../platform/common/utils';
+import { IInterpreterService } from '../../platform/interpreter/contracts';
+import { traceInfoIfCI, traceVerbose, traceDecoratorVerbose, traceError } from '../../platform/logging';
+import { createActiveInterpreterController } from './helpers';
+import {
+    IControllerDefaultService,
+    IControllerLoader,
+    IControllerRegistration,
+    IVSCodeNotebookController
+} from './types';
+
+@injectable()
+export class ControllerDefaultService implements IControllerDefaultService {
+    private get isLocalLaunch(): boolean {
+        return this.serverConnectionType.isLocalLaunch;
+    }
+    constructor(
+        @inject(IControllerRegistration) private readonly registration: IControllerRegistration,
+        @inject(IControllerLoader) private readonly loader: IControllerLoader,
+        @inject(IInterpreterService) private readonly interpreters: IInterpreterService,
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) readonly disposables: IDisposableRegistry,
+        @inject(ServerConnectionType) private readonly serverConnectionType: ServerConnectionType,
+        @inject(IsWebExtension) private readonly isWeb: boolean
+    ) {}
+    public async computeDefaultController(
+        resource: Resource,
+        viewType: typeof JupyterNotebookView | typeof InteractiveWindowView
+    ): Promise<IVSCodeNotebookController | undefined> {
+        if (this.isLocalLaunch) {
+            traceInfoIfCI('CreateActiveInterpreterController');
+            return createActiveInterpreterController(viewType, resource, this.interpreters, this.registration);
+        } else {
+            traceInfoIfCI('CreateDefaultRemoteController');
+            const notebook =
+                viewType === JupyterNotebookView
+                    ? this.notebook.notebookDocuments.find((item) => item.notebookType === viewType)
+                    : undefined;
+            const controller = await this.createDefaultRemoteController(viewType, notebook);
+            // If we're running on web, there is no active interpreter to fall back to
+            if (controller || this.isWeb) {
+                return controller;
+            }
+            traceVerbose('No default remote controller, hence returning the active interpreter');
+            return createActiveInterpreterController(viewType, resource, this.interpreters, this.registration);
+        }
+    }
+
+    @traceDecoratorVerbose('Get default Remote Controller')
+    private async createDefaultRemoteController(
+        notebookType: typeof JupyterNotebookView | typeof InteractiveWindowView,
+        notebook?: NotebookDocument
```

Comment:
I don't think we need the first argument, the notebook type is a property of NotebookDocument object.

Reply:
Yeah if there's a notebook document. In the case of the IW, there isn't one yet.

Reply:
Here's that same method in the original file:

```typescript
    public async getActiveInterpreterOrDefaultController(
        notebookType: typeof JupyterNotebookView | typeof InteractiveWindowView,
        resource: Resource
    ): Promise<IVSCodeNotebookController | undefined> {
        if (this.isLocalLaunch) {
            traceInfoIfCI('CreateActiveInterpreterController');
            return this.createActiveInterpreterController(notebookType, resource);
        } else {
            traceInfoIfCI('CreateDefaultRemoteController');
            const notebook =
                notebookType === JupyterNotebookView
                    ? this.notebook.notebookDocuments.find((item) => item.notebookType === notebookType)
                    : undefined;
            const controller = await this.createDefaultRemoteController(notebookType, notebook);
            // If we're running on web, there is no active interpreter to fall back to
            if (controller || this.isWeb) {
                return controller;
            }
            traceVerbose('No default remote controller, hence returning the active interpreter');
            return this.createActiveInterpreterController(notebookType, resource);
        }
    }

```